sdk/core/azure-core-amqp/src/models/amqp_properties.cpp

```diff
@@ -0,0 +1,442 @@
+        && CompareNullable(CorrelationId, that.CorrelationId)
+        && CompareNullable(UserId, that.UserId) && CompareNullable(To, that.To)
+        && CompareNullable(Subject, that.Subject) && CompareNullable(ReplyTo, that.ReplyTo)
+        && CompareNullable(ContentType, that.ContentType)
+        && CompareNullable(ContentEncoding, that.ContentEncoding)
+        && CompareNullable(AbsoluteExpiryTime, that.AbsoluteExpiryTime)
+        && CompareNullable(CreationTime, that.CreationTime)
+        && CompareNullable(GroupId, that.GroupId)
+        && CompareNullable(GroupSequence, that.GroupSequence)
+        && CompareNullable(ReplyToGroupId, that.ReplyToGroupId));
+  }
+
+  bool MessageProperties::ShouldSerialize() const noexcept
+  {
+    return (
+        MessageId.HasValue() || CorrelationId.HasValue() || UserId.HasValue() || To.HasValue()
+        || Subject.HasValue() || ReplyTo.HasValue() || ContentType.HasValue()
+        || ContentEncoding.HasValue() || AbsoluteExpiryTime.HasValue() || CreationTime.HasValue()
+        || GroupId.HasValue() || GroupSequence.HasValue() || ReplyToGroupId.HasValue());
+  }
+
+  size_t MessageProperties::GetSerializedSize(MessageProperties const& properties)
+  {
+    auto handle = _internal::MessagePropertiesFactory::ToUamqp(properties);
+    AmqpValue propertiesAsValue{amqpvalue_create_properties(handle.get())};
+    return AmqpValue::GetSerializedSize(propertiesAsValue);
+  }
+
+  std::vector<uint8_t> MessageProperties::Serialize(MessageProperties const& properties)
+  {
+    auto handle = _internal::MessagePropertiesFactory::ToUamqp(properties);
+    AmqpValue propertiesAsValue{amqpvalue_create_properties(handle.get())};
+    return Azure::Core::Amqp::Models::AmqpValue::Serialize(propertiesAsValue);
+  }
+
+  MessageProperties MessageProperties::Deserialize(uint8_t const* data, size_t size)
+  {
+    AmqpValue value{AmqpValue::Deserialize(data, size)};
+    PROPERTIES_HANDLE handle;
+    if (amqpvalue_get_properties(value, &handle))
+    {
+      throw std::runtime_error("Could not convert value to AMQP Properties.");
+    }
+    UniquePropertiesHandle uniqueHandle{handle};
+    handle = nullptr;
+    return _internal::MessagePropertiesFactory::FromUamqp(uniqueHandle);
+  }
+
+  namespace {
+    std::string timeToString(std::chrono::system_clock::time_point t)
+    {
+      std::time_t time = std::chrono::system_clock::to_time_t(t);
+      std::string time_str = std::ctime(&time);
+      time_str.resize(time_str.size() - 1);
+      return time_str;
+    }
+
+    size_t LogRawData(std::ostream& os, size_t startOffset, const uint8_t* const pb, size_t cb)
+    {
+      // scratch buffer which will hold the data being logged.
+      std::stringstream ss;
+
+      size_t bytesToWrite = (cb < 0x10 ? cb : 0x10);
+
+      ss << std::hex << std::right << std::setw(8) << std::setfill('0') << startOffset << ": ";
+
+      // Write the buffer data out.
+      for (size_t i = 0; i < bytesToWrite; i += 1)
```

Comment:
If I ever get compromised and I need to quietly signal you that I cannot be trusted, somewhere in my code I'll write `i += 1` or `i++` instead of `++i` :)

Reply:
Yeah - that's a personal idiosyncrasy. For some reason, I always prefer writing += 1 instead of ++, especially in for loops.

And I have absolutely no rational explanation for it (other than I find ++i and i++ distasteful).

Reply:
`++i` allows for micro-optimiation for custom overloaded prefix form of `operator++()` that returns a reference instead of copy. besides that, not all types have `operator+(int)`, so it is more universal - you can write it for iterators that are not random access iterators.

Reply:
Yup - as I mentioned, it's a personal idiosyncrasy. 

I honestly can't really explain it, but it's a thing.